iroha_introspect/iroha_introspect_derive/src/lib.rs

```diff
@@ -0,0 +1,338 @@
+//! Crate with derive Introspect macro
+
+#![allow(
+    clippy::unwrap_used,
+    clippy::expect_used,
+    clippy::panic,
+    clippy::clippy::unimplemented
+)]
+
+use proc_macro::TokenStream;
+use proc_macro2::Ident;
+use proc_macro2::Span;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+use syn::{
+    parse::Parse, parse_macro_input, spanned::Spanned, Attribute, Data, DataEnum, DataStruct,
+    DeriveInput, Expr, Field, Fields, FieldsNamed, FieldsUnnamed, GenericParam, Generics, Lit,
+    LitStr, Meta, NestedMeta, Type, TypePath, Variant,
+};
+
+/// Check out docs in `iroha_introspect` crate
+#[proc_macro_derive(Introspect)]
+pub fn introspect_derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    impl_introspect(&input).into()
+}
+
+fn impl_introspect(input: &DeriveInput) -> TokenStream2 {
+    let name = &input.ident;
+    let (params, ident_params, where_clause) = generics(&input.generics);
+    let type_name: TypePath = syn::parse2(quote! { #name <#(#ident_params),*> }).unwrap();
+    let metadata = metadata(&input.data);
+
+    let type_name_body = type_name_body(name, &input.generics);
+
+    quote! {
+        impl #params iroha_introspect::Introspect for #type_name
+        #where_clause
+        {
+            fn type_name() -> String {
+                #type_name_body
+            }
+
+            fn introspect(map: &mut iroha_introspect::MetaMap) {
+               #metadata
+            }
+        }
+    }
+}
+
+fn type_name_body(name: &Ident, generics: &Generics) -> TokenStream2 {
```

Comment:
Please add docs to all functions here. It is difficult to understand macro generation code usually so function level docs will help a lot